File: server/src/controllers/settingsController.js
import { Settings } from "../models/Settings.js";

export const getSettings = async (req, res) => {
  try {
    const userId = req.userId;
    if (!userId) {
      return res.status(401).json({ message: "Unauthorized" });
    }

    let settings = await Settings.findOne({ user: userId }).lean();

    if (!settings) {
      const created = await Settings.create({ user: userId });
      settings = created.toObject();
    }

    return res.status(200).json(settings);
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error("Get settings error", error);
    return res
      .status(500)
      .json({ message: "Failed to fetch settings", error: String(error) });
  }
};

export const updateSettings = async (req, res) => {
  try {
    const userId = req.userId;
    if (!userId) {
      return res.status(401).json({ message: "Unauthorized" });
    }

    const { dailyGoalMinutes, dailyGoalSessions, goalType, focusArea, showMotivation } = req.body;

    // Build update object with only provided fields
    const updateData = {};
    if (typeof dailyGoalMinutes === "number") updateData.dailyGoalMinutes = dailyGoalMinutes;
    if (typeof dailyGoalSessions === "number") updateData.dailyGoalSessions = dailyGoalSessions;
    if (goalType && (goalType === "sessions" || goalType === "minutes")) updateData.goalType = goalType;
    if (focusArea) updateData.focusArea = focusArea;
    if (typeof showMotivation === "boolean") updateData.showMotivation = showMotivation;

    const updated = await Settings.findOneAndUpdate(
      { user: userId },
      updateData,
      { new: true, upsert: true }
    ).lean();

    return res.status(200).json(updated);
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error("Update settings error", error);
    return res
      .status(500)
      .json({ message: "Failed to update settings", error: String(error) });
  }
};

File: backend/src/controllers/settings_controller.py
from typing import Optional

from pydantic import BaseModel

from ..models import settings as settings_model


class UpdateSettingsBody(BaseModel):
    dailyGoalMinutes: Optional[float] = None
    dailyGoalSessions: Optional[float] = None
    goalType: Optional[str] = None
    focusArea: Optional[str] = None
    showMotivation: Optional[bool] = None


async def get_settings(user_id: str) -> dict:
    doc = await settings_model.find_by_user(user_id)
    if not doc:
        doc = await settings_model.create_default(user_id)
    return settings_model.serialize(doc)


async def update_settings(user_id: str, body: UpdateSettingsBody) -> dict:
    update: dict = {}
    if isinstance(body.dailyGoalMinutes, (int, float)):
        update["dailyGoalMinutes"] = body.dailyGoalMinutes
    if isinstance(body.dailyGoalSessions, (int, float)):
        update["dailyGoalSessions"] = body.dailyGoalSessions
    if body.goalType in ("sessions", "minutes"):
        update["goalType"] = body.goalType
    if body.focusArea:
        update["focusArea"] = body.focusArea
    if isinstance(body.showMotivation, bool):
        update["showMotivation"] = body.showMotivation

    doc = await settings_model.upsert(user_id, update)
    return settings_model.serialize(doc)
